File: ruby/test/unit/config_loader_test.rb
# frozen_string_literal: true

require_relative "../test_helper"

class ConfigLoaderTest < Minitest::Test
  def test_parse_driver_config_string
    json = <<~JSON
      {"driver_id": "redis-rb", "mode": "standalone"}
    JSON

    config = RespBench::Config::ConfigLoader.parse_driver_config_string(json)

    assert_equal "redis-rb", config.driver_id
    assert_equal "standalone", config.mode
  end

  def test_parse_driver_config_with_all_fields
    json = {
      schema_version: "1.0",
      description: "Test driver",
      driver_id: "redis-rb",
      mode: "cluster",
      tls: { enabled: true },
      auth: { password: "secret" },
      specific_driver_config: { timeout: 5000 }
    }

    config = RespBench::Config::ConfigLoader.parse_driver_config(json)

    assert_equal "1.0", config.schema_version
    assert_equal "Test driver", config.description
    assert_equal "redis-rb", config.driver_id
    assert_equal "cluster", config.mode
    assert config.tls
    assert config.auth
    assert_equal 5000, config.specific_driver_config[:timeout]
  end

  def test_parse_workload_config
    json = {
      schema_version: "1.0",
      benchmark_profile: { name: "Test Benchmark", description: "A test" },
      phases: [
        {
          id: "WARMUP",
          connections: 10,
          completion: { type: "requests", requests: 1000 },
          keyspace: {
            keys_count: 1000,
            key_prefix: "bench:",
            generation_alg: "sequential_int"
          },
          commands: [
            { command: "set", weight: 1.0, data_size_bytes: 256 }
          ]
        }
      ]
    }

    config = RespBench::Config::ConfigLoader.parse_workload_config(json)

    assert_equal "Test Benchmark", config.name
    assert_equal 1, config.phases.size

    phase = config.phases.first
    assert_equal "WARMUP", phase.id
    assert_equal 10, phase.connections
    assert phase.completion.request_based?
    assert_equal 1000, phase.completion.total_requests
    assert_equal 1000, phase.keyspace.keys_count
    assert phase.keyspace.sequential_int?
    assert_equal 1, phase.commands.size
  end

  def test_parse_duration_completion
    json = {
      schema_version: "1.0",
      benchmark_profile: { name: "Test" },
      phases: [
        {
          id: "STEADY",
          connections: 1,
          completion: { type: "duration", seconds: 60 },
          keyspace: { keys_count: 100, key_prefix: "test:", generation_alg: "sequential_int" },
          commands: [{ command: "get", weight: 1.0 }]
        }
      ]
    }

    config = RespBench::Config::ConfigLoader.parse_workload_config(json)
    phase = config.phases.first

    assert phase.completion.duration_based?
    assert_equal 60, phase.completion.duration_seconds
  end

  def test_parse_phase_with_limits
    json = {
      schema_version: "1.0",
      benchmark_profile: { name: "Test" },
      phases: [
        {
          id: "LIMITED",
          connections: 10,
          cps_limit: 5,
          rps_limit: 1000,
          pipeline_depth: 4,
          warmup_requests: 2,
          completion: { type: "requests", requests: 1000 },
          keyspace: { keys_count: 100, key_prefix: "test:", generation_alg: "sequential_int" },
          commands: [{ command: "get", weight: 1.0 }]
        }
      ]
    }

    config = RespBench::Config::ConfigLoader.parse_workload_config(json)
    phase = config.phases.first

    assert phase.cps_limit?
    assert_equal 5, phase.cps_limit
    assert phase.rps_limit?
    assert_equal 1000, phase.rps_limit
    assert_equal 4, phase.pipeline_depth
    assert_equal 2, phase.warmup_requests
  end

  def test_parse_uniform_rand_keyspace
    json = {
      schema_version: "1.0",
      benchmark_profile: { name: "Test" },
      phases: [
        {
          id: "RANDOM",
          connections: 1,
          completion: { type: "requests", requests: 100 },
          keyspace: {
            keys_count: 10_000,
            key_prefix: "rand:",
            generation_alg: "uniform_rand",
            seed: 12345
          },
          commands: [{ command: "get", weight: 1.0 }]
        }
      ]
    }

    config = RespBench::Config::ConfigLoader.parse_workload_config(json)
    phase = config.phases.first

    assert phase.keyspace.uniform_rand?
    assert_equal 12345, phase.keyspace.seed
  end

  def test_load_driver_config_from_file
    # Create a temporary config file
    require "tempfile"
    file = Tempfile.new(["driver", ".json"])
    file.write('{"driver_id": "redis-rb", "mode": "standalone"}')
    file.close

    config = RespBench::Config::ConfigLoader.load_driver_config(file.path)

    assert_equal "redis-rb", config.driver_id
  ensure
    file&.unlink
  end

  # Parity with Java's ConfigLoaderTest.shouldDetectClusterMode
  def test_detect_cluster_mode
    config = RespBench::Config::ConfigLoader.parse_driver_config_string(
      '{"driver_id": "redis-rb", "mode": "cluster"}'
    )

    assert config.cluster?
    refute config.standalone?
  end

  # Parity with Java's ConfigLoaderTest.shouldDetectTlsEnabled
  def test_detect_tls_enabled
    config = RespBench::Config::ConfigLoader.parse_driver_config_string(
      '{"driver_id": "redis-rb", "mode": "standalone", "tls": {}}'
    )

    refute_nil config.tls, "TLS config should be present"
  end

  # Parity with Java's ConfigLoaderTest.shouldThrowOnMissingDriverId
  # Note: Ruby's ConfigLoader doesn't validate missing fields (creates a config with nil driver_id).
  # This test documents that behavior. If validation is added later, update accordingly.
  def test_missing_driver_id_results_in_nil
    config = RespBench::Config::ConfigLoader.parse_driver_config_string('{}')

    assert_nil config.driver_id
  end
end

File: node/test/unit/configLoader.test.ts
import assert from 'node:assert/strict';
import { mkdtempSync, writeFileSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import { describe, it } from 'node:test';

import { DEFAULT_DATA_SIZE_BYTES } from '../../src/config/commandConfig.js';
import {
  DEFAULT_KEY_PREFIX,
  DEFAULT_KEY_SIZE_BYTES,
} from '../../src/config/keyspaceConfig.js';
import { ConfigError, ConfigLoader } from '../../src/config/loader.js';

function writeTempJson(name: string, data: unknown): string {
  const dir = mkdtempSync(join(tmpdir(), 'resp-bench-node-'));
  const path = join(dir, name);
  writeFileSync(path, JSON.stringify(data), 'utf8');
  return path;
}

const MINIMAL_PHASE = {
  id: 'STEADY',
  connections: 2,
  completion: { type: 'requests', requests: 100 },
  keyspace: { keys_count: 50 },
  commands: [{ command: 'get', weight: 1.0 }],
};

describe('ConfigLoader — driver config', () => {
  it('parses every schema field', () => {
    const config = ConfigLoader.parseDriverConfig({
      schema_version: '1.0',
      description: 'test driver',
      driver_id: 'ioredis',
      mode: 'cluster',
      command_timeout_ms: 10_000,
      tls: { enabled: true, ca_path: '/tmp/ca.pem' },
      auth: { username: 'u', password: 'p' },
      specific_driver_config: { secondary_driver_id: 'other' },
    });

    assert.equal(config.schemaVersion, '1.0');
    assert.equal(config.description, 'test driver');
    assert.equal(config.driverId, 'ioredis');
    assert.equal(config.mode, 'cluster');
    assert.equal(config.commandTimeoutMs, 10_000);
    assert.equal(config.tlsEnabled(), true);
    assert.equal(config.hasAuth(), true);
    assert.equal(config.isCluster(), true);
    assert.equal(config.isStandalone(), false);
    assert.equal(config.secondaryDriverId(), 'other');
  });

  it('applies defaults for the optional fields', () => {
    const config = ConfigLoader.parseDriverConfig({ driver_id: 'ioredis' });
    assert.equal(config.schemaVersion, '1.0');
    assert.equal(config.mode, 'standalone');
    assert.equal(config.commandTimeoutMs, null);
    assert.equal(config.tlsEnabled(), false);
    assert.equal(config.hasAuth(), false);
    assert.deepEqual(config.specificDriverConfig, {});
    assert.equal(config.secondaryDriverId(), null);
  });

  it('treats an empty username and password as no auth', () => {
    const config = ConfigLoader.parseDriverConfig({
      driver_id: 'ioredis',
      auth: { username: '', password: '' },
    });
    assert.equal(config.hasAuth(), false);
  });

  it('rejects a missing driver_id', () => {
    assert.throws(() => ConfigLoader.parseDriverConfig({ mode: 'standalone' }), ConfigError);
  });

  it('rejects an unknown mode', () => {
    assert.throws(
      () => ConfigLoader.parseDriverConfig({ driver_id: 'ioredis', mode: 'galaxy' }),
      /must be standalone, cluster or sentinel/,
    );
  });

  it('loads the real repo driver configs', () => {
    const path = writeTempJson('driver.json', {
      schema_version: '1.0',
      description: 'valkey-glide-node client - default configuration',
      driver_id: 'valkey-glide-node',
      mode: 'standalone',
      specific_driver_config: {},
    });
    assert.equal(ConfigLoader.loadDriverConfig(path).driverId, 'valkey-glide-node');
  });

  it('reports a readable error for a missing file', () => {
    assert.throws(() => ConfigLoader.loadDriverConfig('/nope/missing.json'), /cannot read driver config/);
  });

  it('reports a readable error for malformed JSON', () => {
    const dir = mkdtempSync(join(tmpdir(), 'resp-bench-node-'));
    const path = join(dir, 'bad.json');
    writeFileSync(path, '{ not json', 'utf8');
    assert.throws(() => ConfigLoader.loadDriverConfig(path), /is not valid JSON/);
  });
});

describe('ConfigLoader — workload config', () => {
  it('parses a full workload', () => {
    const workload = ConfigLoader.parseWorkloadConfig({
      schema_version: '1.0',
      benchmark_profile: { name: 'Reference', description: 'd', version: '1.0.0' },
      phases: [
        {
          id: 'WARMUP',
          description: 'populate',
          connections: 1,
          cps_limit: -1,
          rps_limit: -1,
          pipeline_depth: 4,
          warmup_requests: 3,
          completion: { type: 'requests', requests: 1_000_000 },
          keyspace: {
            keys_count: 1_000_000,
            key_size_bytes: 16,
            key_prefix: 'bench:',
            generation_alg: 'uniform_rand',
            seed: 12345,
          },
          commands: [
            { command: 'get', weight: 0.8 },
            { command: 'set', weight: 0.2, data_size_bytes: 512 },
          ],
        },
      ],
    });

    assert.equal(workload.name(), 'Reference');
    assert.equal(workload.phases.length, 1);
    const phase = workload.phases[0]!;
    assert.equal(phase.id, 'WARMUP');
    assert.equal(phase.connections, 1);
    assert.equal(phase.pipelineDepth, 4);
    assert.equal(phase.effectivePipelineDepth(), 4);
    assert.equal(phase.warmupRequests, 3);
    assert.equal(phase.hasCpsLimit(), false);
    assert.equal(phase.hasRpsLimit(), false);
    assert.equal(phase.completion.isRequestBased(), true);
    assert.equal(phase.completion.totalRequests(), 1_000_000);
    assert.equal(phase.keyspace.isUniformRand(), true);
    assert.equal(phase.keyspace.seedValue(), 12345);
    assert.equal(phase.commands[1]!.dataSizeBytes, 512);
  });

  it('applies the cross-engine defaults', () => {
    const workload = ConfigLoader.parseWorkloadConfig({ phases: [MINIMAL_PHASE] });
    const phase = workload.phases[0]!;
    assert.equal(phase.cpsLimit, -1);
    assert.equal(phase.rpsLimit, -1);
    assert.equal(phase.pipelineDepth, 1);
    assert.equal(phase.warmupRequests, 1);
    assert.equal(phase.description, null);
    assert.equal(phase.keyspace.keySizeBytes, DEFAULT_KEY_SIZE_BYTES);
    assert.equal(phase.keyspace.keyPrefix, DEFAULT_KEY_PREFIX);
    assert.equal(phase.keyspace.isSequentialInt(), true);
    assert.equal(phase.keyspace.seedValue(), 0);
    assert.equal(phase.commands[0]!.dataSizeBytes, DEFAULT_DATA_SIZE_BYTES);
    assert.equal(phase.commands[0]!.weight, 1.0);
    assert.equal(workload.name(), 'unnamed');
  });

  it('treats an explicit null as absent, not as null', () => {
    // The Ruby/Python engines coerce nulls to defaults; a null leaking through
    // would surface as NaN padding or a null prefix deep in the worker loop.
    const workload = ConfigLoader.parseWorkloadConfig({
      phases: [
        {
          ...MINIMAL_PHASE,
          cps_limit: null,
          rps_limit: null,
          pipeline_depth: null,
          warmup_requests: null,
          keyspace: { keys_count: 50, key_size_bytes: null, key_prefix: null, generation_alg: null },
          commands: [{ command: 'set', weight: null, data_size_bytes: null }],
        },
      ],
    });
    const phase = workload.phases[0]!;
    assert.equal(phase.cpsLimit, -1);
    assert.equal(phase.pipelineDepth, 1);
    assert.equal(phase.warmupRequests, 1);
    assert.equal(phase.keyspace.keySizeBytes, DEFAULT_KEY_SIZE_BYTES);
    assert.equal(phase.keyspace.keyPrefix, DEFAULT_KEY_PREFIX);
    assert.equal(phase.keyspace.generationAlg, 'sequential_int');
    assert.equal(phase.commands[0]!.weight, 1.0);
    assert.equal(phase.commands[0]!.dataSizeBytes, DEFAULT_DATA_SIZE_BYTES);
  });

  it('lower-cases command names', () => {
    const workload = ConfigLoader.parseWorkloadConfig({
      phases: [{ ...MINIMAL_PHASE, commands: [{ command: 'GET', weight: 1 }] }],
    });
    assert.equal(workload.phases[0]!.commands[0]!.command, 'get');
  });

  it('recognises limits when set', () => {
    const workload = ConfigLoader.parseWorkloadConfig({
      phases: [{ ...MINIMAL_PHASE, cps_limit: 10, rps_limit: 500 }],
    });
    const phase = workload.phases[0]!;
    assert.equal(phase.hasCpsLimit(), true);
    assert.equal(phase.hasRpsLimit(), true);
  });

  it('rejects a workload with no phases', () => {
    assert.throws(() => ConfigLoader.parseWorkloadConfig({ phases: [] }), /non-empty "phases"/);
    assert.throws(() => ConfigLoader.parseWorkloadConfig({}), /non-empty "phases"/);
  });

  it('rejects a phase with no commands', () => {
    assert.throws(
      () => ConfigLoader.parseWorkloadConfig({ phases: [{ ...MINIMAL_PHASE, commands: [] }] }),
      /non-empty "commands"/,
    );
  });

  it('rejects non-positive connections', () => {
    assert.throws(
      () => ConfigLoader.parseWorkloadConfig({ phases: [{ ...MINIMAL_PHASE, connections: 0 }] }),
      /"connections" must be positive/,
    );
  });

  it('rejects a completion type that carries no target', () => {
    assert.throws(
      () =>
        ConfigLoader.parseWorkloadConfig({
          phases: [{ ...MINIMAL_PHASE, completion: { type: 'duration' } }],
        }),
      /requires a positive "seconds"/,
    );
    assert.throws(
      () =>
        ConfigLoader.parseWorkloadConfig({
          phases: [{ ...MINIMAL_PHASE, completion: { type: 'requests' } }],
        }),
      /requires a positive "requests"/,
    );
  });

  it('rejects an unknown generation_alg', () => {
    assert.throws(
      () =>
        ConfigLoader.parseWorkloadConfig({
          phases: [{ ...MINIMAL_PHASE, keyspace: { keys_count: 10, generation_alg: 'zipf' } }],
        }),
      /sequential_int or uniform_rand/,
    );
  });

  it('rejects a weight outside 0..1, as Java does', () => {
    assert.throws(
      () =>
        ConfigLoader.parseWorkloadConfig({
          phases: [{ ...MINIMAL_PHASE, commands: [{ command: 'get', weight: 5 }] }],
        }),
      /"weight" must be between 0 and 1/,
    );
  });
});
